testing/testing.go

```diff
@@ -113,7 +113,11 @@ func GenSwarm(t *testing.T, opts ...Option) *swarm.Swarm {
 	ps.AddPrivKey(p.ID, p.PrivKey)
 	t.Cleanup(func() { ps.Close() })
 
-	s := swarm.NewSwarm(p.ID, ps, metrics.NewBandwidthCounter(), cfg.connectionGater)
+	var swarmOpts []swarm.Option
+	if cfg.connectionGater != nil {
+		swarmOpts = append(swarmOpts, swarm.WithConnectionGater(cfg.connectionGater))
```

Comment:
I'd change the test swarm config to store a set of swarm "options" instead of a `connectionGater`. We'd then have a `WithSwarmOption` (or something like that) option to pass swarm options directly.

Reply:
I considered that, but unfortunately the upgrader also requires us to set the connection gater.

Reply:
Ah...